nova_vm/src/ecmascript/types/language/bigint.rs

```diff
@@ -232,35 +232,144 @@ impl BigInt {
-                    })
+                    }))
                 }
             }
             (BigInt::SmallBigInt(x), BigInt::BigInt(y))
             | (BigInt::BigInt(y), BigInt::SmallBigInt(x)) => {
                 let x = x.into_i64();
                 let y = &agent[y];
-                agent.heap.create(BigIntHeapData { data: x * &y.data })
+                Ok(agent.heap.create(BigIntHeapData { data: x * &y.data }))
             }
             (BigInt::BigInt(x), BigInt::BigInt(y)) => {
                 let (x, y) = (&agent[x], &agent[y]);
-                agent.heap.create(BigIntHeapData {
+                Ok(agent.heap.create(BigIntHeapData {
                     data: &x.data * &y.data,
-                })
+                }))
+            }
+        }
+    }
+
+    /// ### [BigInt::add ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-add)
+    pub(crate) fn add(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x + y;
+
+                if let Ok(result) = SmallInteger::try_from(result) {
+                    Ok(BigInt::SmallBigInt(SmallBigInt(result)))
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y))
+            | (BigInt::BigInt(y), BigInt::SmallBigInt(x)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x + &y.data }))
+            }
+            (BigInt::BigInt(x), BigInt::BigInt(y)) => {
+                let (x, y) = (&agent[x], &agent[y]);
+                Ok(agent.heap.create(BigIntHeapData {
+                    data: &x.data + &y.data,
+                }))
+            }
+        }
+    }
+
+    /// ### [6.1.6.2.8 BigInt::subtract ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-subtract)
+    pub(crate) fn subtract(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x - y;
+
+                if let Ok(result) = SmallInteger::try_from(result) {
+                    Ok(BigInt::SmallBigInt(SmallBigInt(result)))
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x - &y.data }))
```

Comment:
**issue:** This result, as well as all others in the remainder of this method, could fit in the `SmallBigInt` range.